csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -238,80 +232,100 @@ PersistentBufferStorageParams getPersistentBufferStorageParams(
-  // no need to move buffers to shared memory, return early.
-  if (buffer_params.regs_buffer_size <= available_regs) {
+  // Needs to use rounded shared memory size to avoid over usage.
+  // key : buffer tv.
+  // val : register size and rounded shared memory size
+  std::unordered_map<TensorView*, std::pair<int64_t, int64_t>>
+      required_size_regs_smem_map;
+  int64_t total_smem_buffer_size = 0;
+  for (auto buffer : buffers) {
+    int64_t buffer_size_regs = scheduler_utils::getPersistentBufferSizeOfTensor(
+        buffer, runtime_info, persistent_buffer_info);
+    int64_t buffer_size_smem = roundUpSharedMemory(
+        buffer_size_regs,
+        dataTypeSize(buffer->getDataType().value()),
+        vectorize_factor,
+        InnerOuterPersistentKernelScheduler::threads_per_block_min,
+        InnerOuterPersistentKernelScheduler::threads_per_block_max,
+        dev_prop->warpSize);
+    required_size_regs_smem_map[buffer] =
+        std::make_pair(buffer_size_regs, buffer_size_smem);
+    total_smem_buffer_size += buffer_size_smem;
+  }
+  buffer_params.smem_buffer_size = total_smem_buffer_size;
+  buffer_params.regs_buffer_size =
+      partialOuterReductionBufferSize(reduction_tvs, runtime_info);
+  if (buffer_params.regs_buffer_size <= available_regs &&
+      buffer_params.smem_buffer_size <= available_smem) {
+    buffer_params.smem_persistent_buffers = buffer_params.project_to_input
+        ? persistent_buffer_info.projectable_buffer_inputs
+        : persistent_buffer_info.persistent_buffers;
     buffer_params.has_enough_regs_and_smem = true;
     return buffer_params;
   }
 
-  // (3) Relocate buffers to shared memory until the buffer size in registers is
-  // within the allowable limit.
-  // (3.1) Sort the candidate persistent buffers
-  const auto buffers = buffer_params.project_to_input
-      ? sortProjectableBufferInputs(
-            persistent_buffer_info.projectable_buffer_inputs,
-            outer_broadcast_tvs)
-      : persistent_buffer_info.persistent_buffers;
+  // Moving outer reduction buffer to shared memory is not considered yet,
+  // set to false if the outer reduction buffer size exceeds the register size.
+  if (buffer_params.regs_buffer_size > available_regs) {
+    buffer_params.has_enough_regs_and_smem = false;
+    return buffer_params;
+  }
 
-  // (3.2) Before this loop, all buffers are in registers.
-  // Try to move buffer from register to shared memroy.
-  // After one buffer is moved to shared memory, the buffer size in registers
-  // and shared memory are updated accordingly. Break if required register and
-  // shared memory are lower than limit or shared memory exceeds the limit.
-  int64_t n_smem_buffer = -1;
+  // (2) Now, shared memory is overused, move some buffers to registers.
+  // (2.1) Sort the candidate persistent buffers
+  if (buffer_params.project_to_input) {
+    buffers = sortProjectableBufferInputs(buffers, outer_broadcast_tvs);
```

Comment:
Could you remind me why sorting is done only when projected to inputs?

Reply:
(1) This condition is not unclear. It should be revised to if(!outer_broadcast_tvs.empty()), making it clear that sorting is only needed when outer broadcast tensors are present, as the sorting depends on whether a buffer is a producer of the outer broadcast tensor.
(2) IIRC, it was used to indicate the presence of an outer broadcast tensor in the fusion. When there is an outer broadcast tensor, this scheduler projects the buffer to the inputs, even if it results in a larger buffer size. 
(3) I double-checked the effect of with or without projecting to inputs for fp32 after adding async copy. The conclusion remains the same: projecting to inputs is faster, even with a larger buffer size (3 fp32 when projecting, versus 2 fp32 when not) unless it leads to segmentation (we can patch this latter).
![image](https://github.com/user-attachments/assets/f9d640cd-c815-4540-be33-2f8d018e161d)